ext/ddtrace_profiling_native_extension/private_vm_api_access.c

```diff
@@ -49,6 +50,92 @@ rb_nativethread_id_t pthread_id_for(VALUE thread) {
   #endif
 }
 
+// Queries if the current thread is the owner of the global VM lock.
+//
+// @ivoanjo: Ruby has a similarly-named `ruby_thread_has_gvl_p` but that API is insufficient for our needs because it can
+// still return `true` even when a thread DOES NOT HAVE the global VM lock.
+// In particular, looking at the implementation, that API assumes that if a thread is not in a "blocking region" then it
+// will have the GVL which is probably true for the situations that API was designed to be called from BUT this assumption
+// does not hold true when calling `ruby_thread_has_gvl_p` from a signal handler. (Because the thread may have lost the
+// GVL due to a scheduler decision, not because it decided to block.)
```

Comment:
This is a good question... I think probably not. For a regular C extension, the "blocking region" approach is completely correct -- if you are outside a blocking region then you will have the GVL and if you aren't then you won't.

But because we're interrupting Ruby VM threads half-way through their regular work and calling that API, we see the "extra" situation.

We could try it and see if upstream is interested BUT I have been trying to be very selective, because there's not a lot of resources, so there's a few things I'll try to get fixed/improved (such as https://bugs.ruby-lang.org/issues/19112, https://bugs.ruby-lang.org/issues/19114, https://bugs.ruby-lang.org/issues/18464, ...) before I think it's worth giving this a shot.

Reply:
I was going to comment on that as well. I think it's still worth bringing it up.

Reply:
Fair enough, y'all convinced me 😄. Opened upstream: https://bugs.ruby-lang.org/issues/19172